### PR TITLE
feat: ldap provider login

### PR DESCRIPTION
**Pull request by stebenz**
Add the logic to configure and use LDAP provider as an external IDP with a dedicated login GUI.

```[tasklist]
### Definition of Ready
- [ ] I am happy with the code
- [ ] Short description of the feature/issue is added in the pr description
- [ ] PR is linked to the corresponding user story
- [ ] Acceptance criteria are met
- [ ] All open todos and follow ups are defined in a new ticket and justified
- [ ] Deviations from the acceptance criteria and design are agreed with the PO and documented.
- [ ] No debug or dead code
- [ ] Critical parts are tested automatically
- [ ] Where possible E2E tests are implemented
- [ ] Documentation/examples are up-to-date
- [ ] All non-functional requirements are met
- [ ] Functionality of the acceptance criteria is checked manually on the dev system.
```

### CODE REVIEW

File: internal/command/org_idp_test.go [hunk 1]
```diff
@@ -1967,55 +1950,13 @@ func TestCommandSide_AddOrgLDAPIDP(t *testing.T) {
 				resourceOwner: "org1",
 				provider: LDAPProvider{
 					Name:   "name",
-					Host:   "host",
 					BaseDN: "baseDN",
 				},
 			},
 			res{
 				err: caos_errors.IsErrorInvalidArgument,
```
Comment (livio-a) [inline internal/command/org_idp_test.go:1957]: please check the new tests, where we check for the error id as well

File: internal/idp/providers/ldap/session.go [hunk 2]
```diff
@@ -14,45 +15,118 @@ import (
...
+			s.Provider.bindPassword,
+			s.Provider.baseDN,
+			s.Provider.getNecessaryAttributes(),
+			s.Provider.userObjectClasses,
+			s.Provider.userFilters,
+			s.User,
+			s.Password,
+			s.Provider.timeout,
+		)
+		if err != nil {
```
Comment (livio-a) [inline internal/idp/providers/ldap/session.go:48]: if multiple servers are configured, but the first server is not reachable, the user is not able to login

Comment (livio-a) [reply internal/idp/providers/ldap/session.go:48]: generally the `try` should only handle connection errors to solve HA scenarios, right?

File: internal/query/idp_template.go [hunk 3]
```diff
@@ -870,15 +878,16 @@ func prepareIDPTemplateByIDQuery(ctx context.Context, db prepareDatabase) (sq.Se
...
+					IDPID:             ldapID.String,
+					Servers:           ldapServers,
+					StartTLS:          ldapStartTls.Bool,
+					BaseDN:            ldapBaseDN.String,
+					BindDN:            ldapBindDN.String,
+					BindPassword:      ldapBindPassword,
+					UserBase:          ldapUserBase.String,
+					UserObjectClasses: ldapUserObjectClasses,
+					UserFilters:       ldapUserFilters,
+					Timeout:           ldapTimeout.Abs(),
```
Comment (livio-a) [inline internal/query/idp_template.go:890]: why do we need an absolute value?

File: internal/api/ui/login/ldap_handler.go [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+	authReq, err := l.getAuthRequestAndParseData(r, data)
+	if err != nil {
+		l.renderError(w, r, authReq, err)
+		return
+	}
+	if data.ResetExternalIDP {
+		userAgentID, _ := http_mw.UserAgentIDFromCtx(r.Context())
+		err := l.authRepo.ResetSelectedIDP(r.Context(), authReq.ID, userAgentID)
+		if err != nil {
+			l.renderError(w, r, authReq, err)
```
Comment (livio-a) [inline internal/api/ui/login/ldap_handler.go:53]:
```suggestion
			l.renderLDAPLogin(w, r, authReq, err)
			return
```

File: internal/api/ui/login/ldap_handler.go [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+		}
+
+		l.handleLoginName(w, r)
+		return
+	}
+
+	identityProvider, err := l.getIDPByID(r, authReq.SelectedIDPConfigID)
+	if err != nil {
+		l.renderError(w, r, authReq, err)
```
Comment (livio-a) [inline internal/api/ui/login/ldap_handler.go:62]:
```suggestion
		l.renderLDAPLogin(w, r, authReq, err)
```

File: internal/api/ui/login/ldap_handler.go [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+		return
+	}
+
+	provider, err := l.ldapProvider(r.Context(), identityProvider)
+	if err != nil {
+		l.renderError(w, r, authReq, err)
```
Comment (livio-a) [inline internal/api/ui/login/ldap_handler.go:68]:
```suggestion
		l.renderLDAPLogin(w, r, authReq, err)
```

**Review by livio-a — CHANGES_REQUESTED** [review]
please also update main and we can should have a look at the events version and unique constraints

File: proto/zitadel/idp.proto [hunk 7]
```diff
@@ -305,15 +306,16 @@ message GoogleConfig {
...
+    repeated string servers = 1;
+    bool start_tls = 2;
+    string base_dn = 3;
+    string bind_dn = 4;
+    string user_base = 5;
+    repeated string user_object_classes = 6;
+    repeated string user_filters = 7;
+    google.protobuf.Duration timeout = 8;
+    LDAPAttributes attributes = 9;
+    Options provider_options = 10;
```
[the remark this answers is not in the document]
Comment (stebenz) [reply proto/zitadel/idp.proto:318]: true, deleted

File: proto/zitadel/management.proto [hunk 8]
```diff
@@ -11262,16 +11263,17 @@ message AddLDAPProviderResponse {
...
-    zitadel.idp.v1.LDAPAttributes attributes = 11;
-    zitadel.idp.v1.Options provider_options = 12;
+    repeated string servers = 3 [(validate.rules).repeated = {max_items: 20, items: {string: {min_len: 1, max_len: 200}}}];
+    bool start_tls = 4;
+    string base_dn = 5 [(validate.rules).string = {min_len: 1, max_len: 200}];
+    string bind_dn = 6 [(validate.rules).string = {min_len: 1, max_len: 200}];
+    string bind_password = 7 [(validate.rules).string = {min_len: 1, max_len: 200}];
+    string user_base = 8 [(validate.rules).string = {min_len: 1, max_len: 200}];
+    repeated string user_object_classes = 9 [(validate.rules).repeated = {max_items: 20, items: {string: {min_len: 1, max_len: 200}}}];
+    repeated string user_filters = 10 [(validate.rules).repeated = {max_items: 20, items: {string: {min_len: 1, max_len: 200}}}];
```
[the remark this answers is not in the document]
Comment (stebenz) [reply proto/zitadel/management.proto:11273]: No, changed with min_items: 1

File: internal/query/projection/idp_template.go [hunk 9]
```diff
@@ -228,14 +229,15 @@ func newIDPTemplateProjection(ctx context.Context, config crdb.StatementHandlerC
...
-			crdb.NewColumn(LDAPUserObjectClassCol, crdb.ColumnTypeText, crdb.Nullable()),
-			crdb.NewColumn(LDAPUserUniqueAttributeCol, crdb.ColumnTypeText, crdb.Nullable()),
-			crdb.NewColumn(LDAPAdminCol, crdb.ColumnTypeText, crdb.Nullable()),
-			crdb.NewColumn(LDAPPasswordCol, crdb.ColumnTypeJSONB, crdb.Nullable()),
+			crdb.NewColumn(LDAPBindDNCol, crdb.ColumnTypeText, crdb.Nullable()),
+			crdb.NewColumn(LDAPBindPasswordCol, crdb.ColumnTypeJSONB, crdb.Nullable()),
+			crdb.NewColumn(LDAPUserBaseCol, crdb.ColumnTypeText, crdb.Nullable()),
+			crdb.NewColumn(LDAPUserObjectClassesCol, crdb.ColumnTypeTextArray, crdb.Nullable()),
+			crdb.NewColumn(LDAPUserFiltersCol, crdb.ColumnTypeTextArray, crdb.Nullable()),
+			crdb.NewColumn(LDAPTimeoutCol, crdb.ColumnTypeInt64, crdb.Nullable()),
```
[the remark this answers is not in the document]
Comment (stebenz) [reply internal/query/projection/idp_template.go:240]: Besides the attributes non should be nullable, changed.

File: internal/api/ui/login/ldap_handler.go [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+	authReq, err := l.getAuthRequestAndParseData(r, data)
+	if err != nil {
+		l.renderError(w, r, authReq, err)
+		return
+	}
+	if data.ResetExternalIDP {
+		userAgentID, _ := http_mw.UserAgentIDFromCtx(r.Context())
+		err := l.authRepo.ResetSelectedIDP(r.Context(), authReq.ID, userAgentID)
+		if err != nil {
+			l.renderLDAPLogin(w, r, authReq, err)
```
Comment (livio-a) [inline internal/api/ui/login/ldap_handler.go:53]:
```suggestion
			l.renderLDAPLogin(w, r, authReq, err)
			return
```

File: internal/idp/providers/ldap/session.go [hunk 11]
```diff
@@ -15,45 +16,119 @@ import (
...
+	defer l.Close()
+
+	if !strings.HasPrefix(server, "ldaps") && startTLS {
+		parts := strings.Split(server, "://")
+		parts = strings.Split(parts[1], ":")
+
+		err = l.StartTLS(&tls.Config{ServerName: parts[0]})
+		if err != nil {
+			return nil, err
+		}
```
Comment (livio-a) [inline internal/idp/providers/ldap/session.go:116]: can we have a short discussion about this function, please?

**Review by livio-a — CHANGES_REQUESTED** [review]
i think a short discussion would solve the last issues, incl. timeout

File: internal/idp/providers/ldap/session.go [hunk 12]
```diff
@@ -30,15 +32,37 @@ type Session struct {
...
+				continue
+			}
+		}
+		// Bind as the admin to search for user
+		err = conn.Bind(s.Provider.bindDN, s.Provider.bindPassword)
+		if err != nil {
+			// Failed bind try with next server
+			continue
+		}
+
```
Comment (livio-a) [inline internal/idp/providers/ldap/session.go]: please outsource that method. that way you're also able to defer correctly.
do we need to handle the parse error differently than the other errors?

File: internal/idp/providers/ldap/session.go [hunk 13]
```diff
@@ -100,39 +118,23 @@ func tryLogin(
 	password string,
 	timeout time.Duration,
```
Comment (livio-a) [inline internal/idp/providers/ldap/session.go:119]: timeout is never used

File: internal/idp/providers/ldap/session.go [hunk 14]
```diff
@@ -47,13 +71,10 @@ func (s *Session) FetchUser(_ context.Context) (idp.User, error) {
...
-			break
+		if err != nil {
+			return nil, err
 		}
 	}
-	if err != nil {
-		return nil, err
-	}
 
 	var emailVerified bool
```
Comment (livio-a) [inline internal/idp/providers/ldap/session.go:79]: i would put the mapping into a separate method, so you're able to test it as well

File: internal/idp/providers/ldap/session.go [hunk 15]
```diff
@@ -142,21 +144,32 @@ func tryLogin(
...
 	return user, nil
 }
 
 func queriesAndToSearchQuery(queries ...string) string {
+	if len(queries) == 0 {
+		return ""
+	}
+	if len(queries) == 1 {
+		return queries[0]
+	}
```
Comment (livio-a) [inline internal/idp/providers/ldap/session.go]: please add some tests for these functions

Comment (livio-a) [reply internal/idp/providers/ldap/session.go]: and i guess there's a better option to build the strings, please at least initialize the used array with a size

**Review by livio-a — CHANGES_REQUESTED** [review]

**Pull request merged**